### DISCUSSION TITLE
Comparison to CuteChess CLI

### DISCUSSION THREAD

**Discussion by ptsouchlos** [Disservin/fastchess#137] [General]
👋 Hello!

This looks like an interesting tool, just curious though if you have done any sort of comparison to the `cutechess-cli`? Mostly I'm asking in terms of speed. Not sure exactly how that would be measured.

**Comment by Disservin** [comment]
Hi,

Thanks for your interest ;) 

I indeed have, mainly because cutechess is laggy with high concurrency and low time controls. 
See https://github.com/cutechess/cutechess/issues/630.  
After finishing the engine communication part for fastchess, vondele tested the posix part and the results are mentioned in the readme.
> Extensively tested for high concurrency (supporting up to 250 threads) with short time controls (0.2+0.002s), it exhibits minimal timeout issues, with only 10 matches out of 20,000 experiencing timeouts.

Now admittedly this test is quite old but the engine communication part is still pretty much the same since then.

Other improvements include:
- an auto generated config json from which you can resume the playing
- faster build time and no qt dependency
- a quick start option
- builtin pentamonial stats tracking

Im always excited about new users or contributors 😉 If you have any issues let me know

**Reply by ptsouchlos** [reply]
Thanks for the reply! 

I've been working on the chess challenge from Sebatian Lague and am now considering writing my own engine with what I've learned, but I'd like to do so in C++. While searching I came across your chess library and this project. I think I'm going to give fast-chess a try as I'm doing SPRT testing already but I currently use `cutechess-cli` for that. 

Is there any interest in adding CMake support to the project?

**Reply by Disservin** [reply]
Oh nice! Engines are fun ;)

I feel that with chess libraries you are always a bit limited. No matter how good it is. It's definitely mature enough for an engine but yea..  I mainly wrote that library because I didnt find the alternatives nice and I needed something to kick start other chess projects in C++.

Also I have to admit that I've never really used CMake that much, but I certainly would be open towards having it in the repo.

If you are in for serious SPRT testing you should consider using https://github.com/AndyGrant/OpenBench as the testing framework.

**Reply by ptsouchlos** [reply]
Yes they most certainly are! I find the algorithms fascinating. 

That's true, but that's the main part that would deter me at the moment: having to write my own `board` implementation. For the time being, I want to focus more on the engine itself. 

Ok I'll see what I can do with regards to CMake support.

Thanks for that. I've come across OpenBench before but haven't dived in to see what it's all about. Once I have a more mature engine I'll look into it more. Thanks!

**Reply by Disservin** [reply]
> That's true, but that's the main part that would deter me at the moment: having to write my own board implementation

I think that played a huge part in the chess challenge,  a lot of people just want to write some search code, without having to worry about the other stuff.

**Reply by ptsouchlos** [reply]
> I think that played a huge part in the chess challenge, a lot of people just want to write some search code, without having to worry about the other stuff.

I think this is very accurate. That's where all the fun is! 😄

**Reply by ptsouchlos** [reply]
Do you have any build docs for fast-chess on windows? Does it require cygwin or MSYS?

**Reply by Disservin** [reply]
The CI provides new builds, https://github.com/Disservin/fast-chess/actions/runs/6150528856, the releases do so too.  

Yeah either one of those would be good if you want to compile it yourself. The builds shouldn’t need msys or cgwin

**Reply by ptsouchlos** [reply]
Ok thanks. I was only asking because I haven't been able to build the project with CMake so far and I wasn't sure if there was something I'm missing. I may file an issue if I can't figure things out

**Reply by Disservin** [reply]
Perhaps because includes are not file relative but root relative? In the Makefile I have 
-Isrc and includes look like this 
#include <path> instead of #include “path“.

**Reply by Disservin** [reply]
I don’t mind changing the directory structure a bit if that helps

**Reply by ptsouchlos** [reply]
I think I've been able to get all the files as part of the binary, it just doesn't build and I'm getting an error I'm not familiar with: 

`C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.37.32822\include\ppltasks.h(2715,52): error C3861: '_ReturnAddress': identifier not found` 

I have to dig into it more to figure out what's going on.

This is the cmakelists so far:

```cmake
cmake_minimum_required(VERSION 3.15 FATAL_ERROR)

project(fast-chess)

set(CMAKE_CXX_STANDARD 17)
set(CMAKE_CXX_STANDARD_REQUIRED ON)
file(GLOB_RECURSE FAST_CHESS_SOURCES CONFIG_DEPENDS "src/*.cpp")
file(GLOB_RECURSE FAST_CHESS_HEADERS CONFIG_DEPENDS "src/*.hpp")

add_executable(${PROJECT_NAME} WIN32 ${FAST_CHESS_SOURCES} ${FAST_CHESS_HEADERS})
target_include_directories(${PROJECT_NAME} PRIVATE src)
target_compile_features(${PROJECT_NAME} PRIVATE cxx_std_17)

target_link_options(${PROJECT_NAME} PRIVATE "/entry:mainCRTStartup")
```

**Reply by Disservin** [reply]
Mhh can you check if the stuff inside the builtin namespace compiles for you? https://github.com/Disservin/chess-library/blob/40658fbbb10f824f273fbaa7ca9b9d720203b01e/src/chess.hpp#L746

**Reply by ptsouchlos** [reply]
I was able to build with Clang on windows but not MSVC. I'll file an issue. 

It would also be helpful to change the directory structure a bit. I can make that part of the PR.
